File: gui/src/scope_gui/frame.py
"""Wire-frame parsing for the ESP32 -> host sample stream.

The format is documented in docs/STREAM.md and written by
esp32/main/stream_frame.c; this module is the read side. Keep all three in
step -- gui/tests/test_c_interop.py checks this module against bytes produced
by the real firmware packer.
"""

from __future__ import annotations

import zlib
from dataclasses import dataclass

MAGIC = 0x53434F50  # "SCOP", little-endian on the wire
VERSION = 1
HDR_BYTES = 36

# Columns beyond this are treated as corrupt framing, not a real capture --
# the firmware only ever sends 1000, this is a generous sanity ceiling.
MAX_COLS = 4096

FLAG_PEAK = 1 << 0
FLAG_TRIGGERED = 1 << 1
FLAG_OVERRANGE = 1 << 2

AFE_DC_COUPLED = 1 << 0
AFE_TERM_50R = 1 << 1
AFE_PREAMP_HG = 1 << 2

# LMH6518 preamp steps (datasheet), and its ladder attenuator step size.
_PREAMP_LG_DB = 18.8
_PREAMP_HG_DB = 38.8
_LADDER_STEP_DB = 2.0

# Input divider network ratios for afe_atten 0/1/2.
_DIVIDER_RATIO = (1.0, 10.0, 100.0)

# AD9215 default internal reference gives a 2.0 Vpp full-scale differential
# input; mid-code (512) is 0 V at the ADC pins. Verify against board
# calibration if measurements drift -- this is the one constant in this file
# that depends on how the reference was actually configured in hardware.
ADC_FS_VOLTS = 2.0
ADC_CODE_MID = 512
_ADC_LSB_VOLTS = ADC_FS_VOLTS / 1024


class ParseError(Exception):
    """Base class for frame parse failures."""


class BadVersion(ParseError):
    def __init__(self, version: int) -> None:
        self.version = version
        super().__init__(f"unsupported frame version {version} (expected {VERSION})")


class BadColumnCount(ParseError):
    def __init__(self, n_cols: int) -> None:
        self.n_cols = n_cols
        super().__init__(f"implausible column count {n_cols}")


class BadCrc(ParseError):
    def __init__(self) -> None:
        super().__init__("CRC mismatch")


@dataclass(frozen=True)
class Frame:
    seq: int
    sample_count: int
    dec_factor: int
    pre_count: int
    post_count: int
    trig_ptr: int
    trig_level: int
    afe_atten: int
    lmh_atten: int
    dac_offset: int
    peak: bool
    triggered: bool
    overrange: bool
    dc_coupled: bool
    term_50r: bool
    preamp_hg: bool
    cols: tuple[tuple[int, int], ...]  # (ymin, ymax) codes, one pair per column

    def sample_period(self) -> float:
        """Seconds per sample, from the 105 MHz encode clock and decimation."""
        return (self.dec_factor + 1) / 105_000_000.0

    def duration(self) -> float:
        return self.sample_count * self.sample_period()

    def volts_per_count(self) -> float:
        """Volts at the probe tip per ADC code step, given the current AFE
        settings (input divider, LMH6518 preamp + ladder attenuation)."""
        preamp_db = _PREAMP_HG_DB if self.preamp_hg else _PREAMP_LG_DB
        gain_db = preamp_db - self.lmh_atten * _LADDER_STEP_DB
        gain_v_per_v = 10 ** (gain_db / 20.0)
        divider = _DIVIDER_RATIO[self.afe_atten]
        return _ADC_LSB_VOLTS * divider / gain_v_per_v

    def code_to_volts(self, code: int) -> float:
        return (code - ADC_CODE_MID) * self.volts_per_count()

    def trig_fraction(self) -> float | None:
        """Trigger position as a fraction of the record, or None if the record
        is empty."""
        if self.sample_count == 0:
            return None
        return min(1.0, self.trig_ptr / self.sample_count)


def cols_of(header: bytes) -> int:
    """Reads n_cols out of a buffer that starts at the magic word and holds at
    least the version and n_cols fields. Raises BadVersion/BadColumnCount
    without needing the CRC or payload, so a caller can size its read before
    the rest of the frame has arrived."""
    version = header[4]
    if version != VERSION:
        raise BadVersion(version)
    n_cols = int.from_bytes(header[6:8], "little")
    if n_cols == 0 or n_cols > MAX_COLS:
        raise BadColumnCount(n_cols)
    return n_cols


def parse(buf: bytes) -> Frame:
    """Parses one complete frame (magic word through the trailing CRC)."""
    n_cols = cols_of(buf)
    payload_end = HDR_BYTES + n_cols * 4
    total = payload_end + 4

    crc_stored = int.from_bytes(buf[payload_end:total], "little")
    crc_calc = zlib.crc32(buf[4:payload_end]) & 0xFFFFFFFF
    if crc_calc != crc_stored:
        raise BadCrc()

    flags = buf[5]
    afe_flags = buf[31]

    cols = tuple(
        (
            int.from_bytes(buf[off:off + 2], "little"),
            int.from_bytes(buf[off + 2:off + 4], "little"),
        )
        for off in range(HDR_BYTES, payload_end, 4)
    )

    return Frame(
        seq=int.from_bytes(buf[8:12], "little"),
        sample_count=int.from_bytes(buf[12:16], "little"),
        dec_factor=int.from_bytes(buf[16:18], "little"),
        pre_count=int.from_bytes(buf[18:20], "little"),
        post_count=int.from_bytes(buf[20:22], "little"),
        trig_ptr=int.from_bytes(buf[22:26], "little"),
        trig_level=int.from_bytes(buf[26:28], "little"),
        afe_atten=buf[30],
        lmh_atten=buf[32],
        dac_offset=int.from_bytes(buf[34:36], "little"),
        peak=bool(flags & FLAG_PEAK),
        triggered=bool(flags & FLAG_TRIGGERED),
        overrange=bool(flags & FLAG_OVERRANGE),
        dc_coupled=bool(afe_flags & AFE_DC_COUPLED),
        term_50r=bool(afe_flags & AFE_TERM_50R),
        preamp_hg=bool(afe_flags & AFE_PREAMP_HG),
        cols=cols,
    )

File: gui/src/frame.rs
//! Parser for the ESP32 sample stream. Wire format lives in `docs/STREAM.md`;
//! the writer is `esp32/main/stream.c`. Keep all three in step.

pub const MAGIC: u32 = 0x5343_4F50; // "SCOP"
pub const VERSION: u8 = 1;
pub const HDR_BYTES: usize = 36;

/// Guards against a corrupt length driving a huge allocation.
pub const MAX_COLS: usize = 8192;

// frame flags
const FLAG_PEAK: u8 = 1 << 0;
const FLAG_TRIGGERED: u8 = 1 << 1;
const FLAG_OVERRANGE: u8 = 1 << 2;

// afe_flags
const AFE_DC_COUPLED: u8 = 1 << 0;
const AFE_TERM_50R: u8 = 1 << 1;
const AFE_PREAMP_HG: u8 = 1 << 2;

/// ADC mid-scale; the AD9215 is 10-bit and mid-scale is 0 V at the input.
pub const CODE_MID: f64 = 512.0;

/// Sample rate after the PL133 clock, before decimation.
pub const SAMPLE_RATE_HZ: f64 = 105_000_000.0;

#[derive(Clone, Debug)]
pub struct Frame {
    pub seq: u32,
    pub peak: bool,
    pub triggered: bool,
    pub overrange: bool,
    pub sample_count: u32,
    pub dec_factor: u16,
    pub pre_count: u16,
    pub post_count: u16,
    pub trig_ptr: u32,
    pub trig_level: u16,
    pub afe_atten: u8,
    pub dc_coupled: bool,
    pub term_50r: bool,
    pub preamp_hg: bool,
    pub lmh_atten: u8,
    pub dac_offset: u16,
    /// Per column, the lowest and highest code seen.
    pub cols: Vec<(u16, u16)>,
}

impl Frame {
    /// Seconds per sample after decimation.
    pub fn sample_period(&self) -> f64 {
        (self.dec_factor as f64 + 1.0) / SAMPLE_RATE_HZ
    }

    /// Full record duration in seconds.
    pub fn duration(&self) -> f64 {
        self.sample_count as f64 * self.sample_period()
    }

    /// Volts per ADC count at the probe tip.
    ///
    /// The ESP32 sends the front-end settings raw, so the dB arithmetic happens
    /// here in floating point rather than being rounded into an integer field.
    /// Chain: input divider, then the LMH6518 preamp, then its 2 dB ladder.
    pub fn volts_per_count(&self) -> f64 {
        let divider = match self.afe_atten {
            0 => 1.0,
            1 => 10.0,
            _ => 100.0,
        };
        let preamp_db = if self.preamp_hg { 38.8 } else { 18.8 };
        let gain_db = preamp_db - 2.0 * self.lmh_atten as f64;
        let gain = 10f64.powf(gain_db / 20.0);

        // AD9215 full scale is 2 Vpp across 1024 codes, referred back through
        // the amplifier chain and the input divider.
        const ADC_FULL_SCALE_V: f64 = 2.0;
        (ADC_FULL_SCALE_V / 1024.0) / gain * divider
    }

    /// Convert a raw code to volts at the probe tip.
    pub fn code_to_volts(&self, code: u16) -> f64 {
        (code as f64 - CODE_MID) * self.volts_per_count()
    }

    /// Where the trigger sits as a fraction across the record, 0.0 to 1.0.
    pub fn trig_fraction(&self) -> Option<f64> {
        if self.sample_count == 0 {
            return None;
        }
        let f = self.trig_ptr as f64 / self.sample_count as f64;
        (0.0..=1.0).contains(&f).then_some(f)
    }
}

fn u16_at(b: &[u8], at: usize) -> u16 {
    u16::from_le_bytes([b[at], b[at + 1]])
}

fn u32_at(b: &[u8], at: usize) -> u32 {
    u32::from_le_bytes([b[at], b[at + 1], b[at + 2], b[at + 3]])
}

/// CRC-32, reflected `0xEDB88320`, matching `crc32()` in `esp32/main/stream.c`.
pub fn crc32(data: &[u8]) -> u32 {
    let mut crc = 0xFFFF_FFFFu32;
    for &byte in data {
        crc ^= byte as u32;
        for _ in 0..8 {
            let mask = (crc & 1).wrapping_neg();
            crc = (crc >> 1) ^ (0xEDB8_8320 & mask);
        }
    }
    !crc
}

#[derive(Debug)]
pub enum ParseError {
    BadVersion(u8),
    BadColumnCount(u16),
    BadCrc,
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ParseError::BadVersion(v) => {
                write!(f, "unsupported frame version {v} (expected {VERSION})")
            }
            ParseError::BadColumnCount(n) => write!(f, "implausible column count {n}"),
            ParseError::BadCrc => write!(f, "CRC mismatch"),
        }
    }
}

/// Number of columns declared by a header, for sizing the payload read.
pub fn cols_of(header: &[u8]) -> Result<usize, ParseError> {
    let version = header[4];
    if version != VERSION {
        return Err(ParseError::BadVersion(version));
    }
    let n = u16_at(header, 6);
    if n == 0 || n as usize > MAX_COLS {
        return Err(ParseError::BadColumnCount(n));
    }
    Ok(n as usize)
}

/// Parse a full frame: the 36-byte header, the column payload, and the CRC.
pub fn parse(buf: &[u8]) -> Result<Frame, ParseError> {
    let n_cols = cols_of(buf)?;
    let payload_end = HDR_BYTES + n_cols * 4;

    let want = u32_at(buf, payload_end);
    if crc32(&buf[4..payload_end]) != want {
        return Err(ParseError::BadCrc);
    }

    let flags = buf[5];
    let afe_flags = buf[31];

    let mut cols = Vec::with_capacity(n_cols);
    for c in 0..n_cols {
        let at = HDR_BYTES + c * 4;
        cols.push((u16_at(buf, at), u16_at(buf, at + 2)));
    }

    Ok(Frame {
        seq: u32_at(buf, 8),
        peak: flags & FLAG_PEAK != 0,
        triggered: flags & FLAG_TRIGGERED != 0,
        overrange: flags & FLAG_OVERRANGE != 0,
        sample_count: u32_at(buf, 12),
        dec_factor: u16_at(buf, 16),
        pre_count: u16_at(buf, 18),
        post_count: u16_at(buf, 20),
        trig_ptr: u32_at(buf, 22),
        trig_level: u16_at(buf, 26),
        afe_atten: buf[30],
        dc_coupled: afe_flags & AFE_DC_COUPLED != 0,
        term_50r: afe_flags & AFE_TERM_50R != 0,
        preamp_hg: afe_flags & AFE_PREAMP_HG != 0,
        lmh_atten: buf[32],
        dac_offset: u16_at(buf, 34),
        cols,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The reference check value for the reflected CRC-32 of "123456789".
    #[test]
    fn crc_matches_reference_vector() {
        assert_eq!(crc32(b"123456789"), 0xCBF4_3926);
    }

    fn build(n_cols: u16) -> Vec<u8> {
        let mut b = vec![0u8; HDR_BYTES + n_cols as usize * 4 + 4];
        b[0..4].copy_from_slice(&MAGIC.to_le_bytes());
        b[4] = VERSION;
        b[5] = FLAG_TRIGGERED;
        b[6..8].copy_from_slice(&n_cols.to_le_bytes());
        b[8..12].copy_from_slice(&7u32.to_le_bytes()); // seq
        b[12..16].copy_from_slice(&(n_cols as u32).to_le_bytes()); // sample_count
        b[22..26].copy_from_slice(&(n_cols as u32 / 2).to_le_bytes()); // trig_ptr
        b[30] = 2; // 100x
        for c in 0..n_cols as usize {
            let at = HDR_BYTES + c * 4;
            b[at..at + 2].copy_from_slice(&(c as u16).to_le_bytes());
            b[at + 2..at + 4].copy_from_slice(&(c as u16 + 1).to_le_bytes());
        }
        let end = HDR_BYTES + n_cols as usize * 4;
        let crc = crc32(&b[4..end]);
        b[end..end + 4].copy_from_slice(&crc.to_le_bytes());
        b
    }

    #[test]
    fn round_trips_a_built_frame() {
        let buf = build(8);
        let f = parse(&buf).expect("parses");
        assert_eq!(f.seq, 7);
        assert!(f.triggered);
        assert!(!f.overrange);
        assert_eq!(f.cols.len(), 8);
        assert_eq!(f.cols[3], (3, 4));
        assert_eq!(f.trig_fraction(), Some(0.5));
    }

    #[test]
    fn rejects_a_corrupted_payload() {
        let mut buf = build(8);
        buf[HDR_BYTES] ^= 0xFF;
        assert!(matches!(parse(&buf), Err(ParseError::BadCrc)));
    }

    #[test]
    fn rejects_a_bad_version() {
        let mut buf = build(4);
        buf[4] = 99;
        assert!(matches!(parse(&buf), Err(ParseError::BadVersion(99))));
    }

    #[test]
    fn rejects_an_absurd_column_count() {
        let mut buf = build(4);
        buf[6..8].copy_from_slice(&u16::MAX.to_le_bytes());
        assert!(matches!(parse(&buf), Err(ParseError::BadColumnCount(_))));
    }

    #[test]
    fn attenuation_scales_volts_per_count() {
        let mut f = parse(&build(4)).unwrap();
        f.afe_atten = 0;
        let at_1x = f.volts_per_count();
        f.afe_atten = 2;
        assert!((f.volts_per_count() - at_1x * 100.0).abs() < 1e-12);
    }
}
